mesop/bin/bin.py

```diff
@@ -84,37 +85,113 @@ def main(argv: Sequence[str]):
+
+def clear_app_modules() -> None:
+  for module in app_modules:
+    # Not every module has been loaded into sys.modules (e.g. the main module)
+    if module in sys.modules:
+      del sys.modules[module]
+
+
+def add_app_module(workspace_dir_path: str, app_module_path: str) -> None:
+  module_name = get_app_module_name(
+    workspace_dir_path=workspace_dir_path, app_module_path=app_module_path
+  )
+  app_modules.add(module_name)
+
+
+def remove_app_module(workspace_dir_path: str, app_module_path: str) -> None:
+  module_name = get_app_module_name(
+    workspace_dir_path=workspace_dir_path, app_module_path=app_module_path
+  )
+  app_modules.remove(module_name)
+
+
+def get_app_module_name(workspace_dir_path: str, app_module_path: str) -> str:
+  relative_path = os.path.relpath(app_module_path, workspace_dir_path)
+  return relative_path.replace(os.sep, ".").rstrip(".py")
+
+
+class ReloadEventHandler(FileSystemEventHandler):
+  def __init__(self, absolute_path: str, workspace_dir_path: str):
+    self.absolute_path = absolute_path
+    self.workspace_dir_path = workspace_dir_path
+
+  def on_any_event(self, event: FileSystemEvent):
+    if not event.is_directory:
       try:
         print("Hot reload: starting...")
         reset_runtime()
-        execute_main_module(absolute_path=absolute_path)
+        execute_main_module(absolute_path=self.absolute_path)
         hot_reload_finished()
         print("Hot reload: finished!")
       except Exception as e:
         logging.log(
           logging.ERROR, "Could not hot reload due to error:", exc_info=e
         )
 
-    time.sleep(0.1)
+  def on_created(self, event: FileSystemEvent):
+    src_path = cast(str, event.src_path)
+    if src_path.endswith(".py"):
+      print(f"Watching new Python module: {event.src_path}")
+      add_app_module(
+        workspace_dir_path=self.workspace_dir_path,
+        app_module_path=event.src_path,
+      )
+
+  def on_deleted(self, event: FileSystemEvent):
+    src_path = cast(str, event.src_path)
+    if src_path.endswith(".py"):
+      print(f"Stopped watching deleted Python module: {event.src_path}")
+      remove_app_module(
+        workspace_dir_path=self.workspace_dir_path,
+        app_module_path=event.src_path,
+      )
+
+
+def fs_watcher(absolute_path: str):
+  """
+  Filesystem watcher using watchdog. Watches for any changes in the specified directory
+  and triggers hot reload on change.
+  """
+  workspace_dir_path = os.path.dirname(absolute_path)
+
+  # Initially track all the files on the file system and then rely on watchdog.
+  for root, dirnames, files in os.walk(workspace_dir_path):
+    # Filter out directories starting with "." because they
+    # can be special directories like .venv directories which
+    # can have lots of Python files that are definitely not application Python modules.
+    dirnames[:] = [d for d in dirnames if not d.startswith(".")]
+
+    for file in files:
+      if file.endswith(".py"):
```

Comment:
[question] How does this handle the `__init__.py` case?

Reply:
great catch - I forgot to handle `__init__.py`